### DISCUSSION TITLE
Using Marp over WSL (Windows Subsystem for Linux)

### DISCUSSION THREAD

**Discussion by yhatt** [marp-team/marp#113] [Frequently Asked]
If you are trying to use Marp through WSL (Windows Subsystem for Linux), you might met some trouble.

I hope this post helps troubleshooting for WSL users, and feedbacks are welcome!

### Summary

- **WSL1**: No actions should need. :raised_hands: 
- **WSL2**: Install a compatible browser to Linux, or turn on [mirrored mode networking].

[mirrored mode networking]: https://learn.microsoft.com/windows/wsl/networking#mirrored-mode-networking

# WSL1

For PDF/PPTX/image conversion, Marp over WSL1 will try to spawn Chrome (or Chromium-flavored browser Marp has supported) _that has installed to Windows._

In the most case, your Windows has pre-installed the compatible Edge browser so **no actions should need.** :raised_hands: 

If you want to use Linux browser for some reason (e.g. render the text with fonts installed in Linux), you should specify a path to the browser binary manually.

# WSL2

WSL2 is working with the virtualized environment unlike WSL1 and Windows and WSL Linux are not sharing local network. So Marp working on Linux cannot connect with the spawned browser in Windows.

## Install a compatible browser into Linux

Follow https://learn.microsoft.com/en-us/windows/wsl/tutorials/gui-apps#install-google-chrome-for-linux to install Google Chrome into your Linux.

After install the browser, Marp just should work! 🙌

## Turn on [mirrored mode networking] (Marp CLI v4 and later)

If you want to use Windows browser when converting, you have to set up WSL to use **[mirrored mode networking]**. Put `%USERPROFILE%\.wslconfig` as following:

```ini
# %USERPROFILE%\.wslconfig
[wsl2]
networkingMode=mirrored
```

To apply changes, you may have to run `wsl --shutdown` and re-launch WSL. After than, Marp CLI on WSL 2 should be able to detect the browser from Windows automatically if there are no compatible browsers on Linux.

<details>
<summary>Tips: Set the path to browser executable explicitly</summary>

#### Marp CLI

Set `--browser-path` option.

```bash
marp --browser-path /usr/bin/chromium-browser --pdf markdown.md
```

#### Marp for VS Code + [WSL extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl)

Open extension settings, and set the path to executable in "Markdown > Marp: Browser Path" (`markdown.marp.browserPath`).

</details>

**Comment by AramLR** [comment]
You don't need to set CHROME_PATh to a browser installed with WSL, you can use /mnt/c to access windows files and then access a browser installed in your computer.

**Reply by yhatt** [reply]
Yes. As I mentioned, the current version of Marp CLI can detect the path of Chrome installed to Linux automatically.

> In WSL2, Marp CLI v1.5.1 and later will try to detect installed Chrome/Chromium in Linux

However, WSL2 cannot use installed the browser to Windows (e.g. Edge). You always have to install the browser into Linux.
